File: MunchkinLevelCounter/app/src/main/java/com/gmail/miloszwasacz/munchkinlevelcounter/MainActivity.java
package com.gmail.miloszwasacz.munchkinlevelcounter;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.FrameLayout;
import com.google.gson.Gson;
import java.util.ArrayList;
import java.util.List;
import java.lang.reflect.Type;
import com.google.gson.reflect.TypeToken;
import android.content.SharedPreferences;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity
{

    FloatingActionButton floatingActionButton;
    PlayerAdapter adapter;
    List<Player> list;
    String SharedPrefs;
    int MaxPlayerLevel;
    int MinLevel;
    List<Game> gameList;
    boolean editMode;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        getSupportActionBar().setTitle("Licznik");

        list = new ArrayList<Player>();

        //Tworzenie domyślnej listy graczy
        if(savedInstanceState == null)
            createDefaultGame(list);

        //Przywracanie stanu poprzedniego listy graczy
        else
        {
            MaxPlayerLevel = savedInstanceState.getInt("MaksymalnyPoziom", getResources().getInteger(R.integer.deafult_rules));
            MinLevel = savedInstanceState.getInt("MinimalnyPoziom", getResources().getInteger(R.integer.deafult_min_level));
            String json = savedInstanceState.getString("ListaGraczy");
            Type listType = new TypeToken<ArrayList<Player>>(){}.getType();
            list = new Gson().fromJson(json, listType);
        }

        setPlayerAdapter();

        //Zmiana trybu guzika: edycja/dodawanie graczy
        floatingActionButton = (FloatingActionButton) findViewById(R.id.floatingActionButton);
        floatingActionButton.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View v)
            {
                if (adapter.editMode)
                {
                    list.add(new Player("Nowy gracz", MinLevel));
                    adapter.notifyItemInserted(list.size() - 1);
                }
                else
                    changeEditMode(R.drawable.ic_baseline_add_white_24dp, true, "Edytuj graczy");
            }
        });
    }

    //RecyclerView i PlayerAdapter
    public void setPlayerAdapter()
    {
        final RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recycler_view);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new PlayerAdapter(list);
        adapter.editMode = editMode;

        //Sprawdzenie czy poziomy graczy są w dozwolonym zakresie
        for (Player element:list)
        {
            if(element.level > MaxPlayerLevel)
                element.level = MaxPlayerLevel;
            else if(element.level < MinLevel)
                element.level = MinLevel;
        }

        adapter.setOnItemClickListener(new PlayerAdapter.OnItemClickListener()
        {
            @Override
            public void onItemClick(final int position)
            {
                final FrameLayout frameLayout = (FrameLayout) getLayoutInflater().inflate(R.layout.player_dialog, null, false);
                final EditText editText = (EditText) frameLayout.findViewById(R.id.editText);
                editText.setText(list.get(position).name);

                //Okienko do edycji poszczególnego gracza
                new AlertDialog.Builder(MainActivity.this)
                        .setTitle("Edytuj gracza")
                        .setPositiveButton("Ok", new DialogInterface.OnClickListener()
                        {
                            @Override
                            public void onClick(DialogInterface dialog, int which)
                            {
                                list.get(position).name = editText.getText().toString();
                                adapter.notifyItemChanged(position);
                            }
                        })
                        .setNegativeButton("Anuluj", null)
                        .setNeutralButton("Usuń", new DialogInterface.OnClickListener()
                        {
                            @Override
                            public void onClick(DialogInterface dialog, int which)
                            {
                                list.remove(position);
                                adapter.notifyItemRemoved(position);
                            }
                        })
                        .setView(frameLayout)
                        .create()
                        .show();
            }

            //Zwiększenie poziomu gracza
            @Override
            public void onAddClick(int position)
            {
                if (list.get(position).level < MaxPlayerLevel)
                {
                    list.get(position).level++;
                    adapter.notifyItemChanged(position);
                }
            }

            //Zmniejszenie poziomu gracza
            @Override
            public void onRemoveClick(int position)
            {
                if (list.get(position).level > MinLevel)
                {
                    list.get(position).level--;
                    adapter.notifyItemChanged(position);
                }
            }

            //Wejście w tryb Kill-O-Meter
            @Override
            public void onFightClick(int position)
            {
                Intent intent = new Intent(MainActivity.this, KillOMeterActivity.class);
                int playerLevel = list.get(position).level;
                String playerName = list.get(position).name;
                int playerPosition = position;
                intent.putExtra("EXTRA_LEVEL", playerLevel);
                intent.putExtra("EXTRA_NAME", playerName);
                intent.putExtra("EXTRA_POSITION", playerPosition);
                intent.putExtra("EXTRA_LIST", serializePlayerList(list));
                intent.putExtra("EXTRA_MAX_LEVEL", MaxPlayerLevel);
                intent.putExtra("EXTRA_MIN_LEVEL", MinLevel);
                startActivityForResult(intent, 1);
            }
        });
        recyclerView.setAdapter(adapter);
    }

    //Włączanie/wyłączanie trybu edycji
    public void changeEditMode(int icon, Boolean enabled, String title)
    {
        getSupportActionBar().setTitle(title);
        floatingActionButton.setImageResource(icon);
        editMode = enabled;
        adapter.editMode = enabled;
        invalidateOptionsMenu();
        adapter.notifyDataSetChanged();
    }

    //Strzałeczka w tył (tryb edycji)
    @Override
    public void onBackPressed()
    {
        if (adapter.editMode)
            changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik");
        else
            super.onBackPressed();
    }

    //Tworzenie guzików na app barze
    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        menu.clear();
        getMenuInflater().inflate(R.menu.toolbar_menu, menu);
        MenuItem mSaveButton = menu.findItem(R.id.action_save);
        MenuItem mLoadButton = menu.findItem(R.id.action_folder);
        MenuItem mClearButton = menu.findItem(R.id.action_clear);
        MenuItem mSettingsButton = menu.findItem(R.id.navigation_settings);
        if(adapter.editMode)
        {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            mSaveButton.setVisible(false);
            mLoadButton.setVisible(false);
            mClearButton.setVisible(false);
            mSettingsButton.setVisible(true);
        }
        else
        {
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);
            mSaveButton.setVisible(true);
            mLoadButton.setVisible(true);
            mClearButton.setVisible(true);
            mSettingsButton.setVisible(false);
        }
        return super.onCreateOptionsMenu(menu);
    }

    //Obsługa guzików na app barze
    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        SharedPreferences prefs = getSharedPreferences(SharedPrefs, MODE_PRIVATE);
        final String listaGier = prefs.getString("ListaGierPrefs", null);
        Type listType = new TypeToken<ArrayList<Game>>(){}.getType();
        gameList = new Gson().fromJson(listaGier, listType);
        if(gameList.isEmpty())
            gameList = null;

        switch (item.getItemId())
        {
            //Zapisanie rozgrywki
            case R.id.action_save:
                final FrameLayout frameLayout = (FrameLayout) getLayoutInflater().inflate(R.layout.player_dialog, null, false);
                final EditText editText = (EditText) frameLayout.findViewById(R.id.editText);

                new AlertDialog.Builder(MainActivity.this)
                        .setTitle("Zapisz rozgrywkę")
                        .setPositiveButton("Ok", new DialogInterface.OnClickListener()
                        {
                            @Override
                            public void onClick(DialogInterface dialog, int which)
                            {
                                if(gameList == null)
                                    gameList = new ArrayList<Game>();

                                if(editText.getText().toString().equals(""))
                                    editText.setText("Gra " + (gameList.size() + 1));

                                gameList.add(new Game(editText.getText().toString(), serializePlayerList(list), MaxPlayerLevel, MinLevel));
                                saveGameListInSharedPreferences(gameList);

                                changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik");
                                Toast.makeText(MainActivity.this, "Zapisano rozgrywkę", Toast.LENGTH_SHORT).show();
                            }
                        })
                        .setNegativeButton("Anuluj", null)
                        .setView(frameLayout)
                        .create()
                        .show();

                return true;

            //Wczytanie ostatniej rozgrywki
            case R.id.action_folder:
                if(gameList == null)
                    gameList = new ArrayList<Game>();
                List<Player> lista = new ArrayList<Player>();
                createDefaultGame(lista);
                gameList.add(0, new Game("default", serializePlayerList(lista), MaxPlayerLevel, MinLevel));

                String[] nameArray = new String[gameList.size()];
                for (int i = 0; i < gameList.size(); i++)
                {
                    nameArray[i] = gameList.get(i).name;
                }

                new AlertDialog.Builder(MainActivity.this)
                        .setTitle("Wczytaj rozgrywkę")
                        .setItems(nameArray, new DialogInterface.OnClickListener()
                        {
                            @Override
                            public void onClick(DialogInterface dialog, int which)
                            {
                                MaxPlayerLevel = gameList.get(which).maxLevel;
                                MinLevel = gameList.get(which).minLevel;
                                String jsonGame = gameList.get(which).content;
                                Type listType = new TypeToken<ArrayList<Player>>(){}.getType();
                                list = new Gson().fromJson(jsonGame, listType);
                                setPlayerAdapter();
                                changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik");

                                Toast.makeText(MainActivity.this, "Wczytano rozgrywkę", Toast.LENGTH_SHORT).show();
                            }
                        })
                        .create()
                        .show();

                return true;

            //Usunięcie gry z listy
            case R.id.action_clear:
                if(gameList == null)
                    Toast.makeText(MainActivity.this, "Brak zapisanych rozgrywek", Toast.LENGTH_LONG).show();
                else
                {
                    String[] nameArrayDelete = new String[gameList.size()];

                    for (int i = 0; i < gameList.size(); i++)
                    {
                        nameArrayDelete[i] = gameList.get(i).name;
                    }

                    new AlertDialog.Builder(MainActivity.this)
                            .setTitle("Usuń rozgrywkę")
                            .setItems(nameArrayDelete, new DialogInterface.OnClickListener()
                            {
                                @Override
                                public void onClick(DialogInterface dialog, int which)
                                {
                                    gameList.remove(which);
                                    saveGameListInSharedPreferences(gameList);

                                    changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik");
                                    Toast.makeText(MainActivity.this, "Usunięto rozgrywkę", Toast.LENGTH_SHORT).show();
                                }
                            })
                            .create()
                            .show();
                }
                return true;

            //Otwarcie ustawień
            case R.id.navigation_settings:
                Intent intent = new Intent(MainActivity.this, SettingsActivity.class);
                intent.putExtra("EXTRA_LIST", serializePlayerList(list));
                intent.putExtra("EXTRA_MAX_LEVEL", MaxPlayerLevel);
                intent.putExtra("EXTRA_MIN_LEVEL", MinLevel);
                intent.putExtra("EXTRA_EDIT_MODE", adapter.editMode);
                startActivityForResult(intent, 2);
                return true;

            //Strzałeczka "back"
            case android.R.id.home:
                onBackPressed();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    //Zapisywanie stanu listy graczy
    @Override
    public void onSaveInstanceState(Bundle savedInstanceState)
    {
        super.onSaveInstanceState(savedInstanceState);
        Gson gson = new Gson();
        String json = gson.toJson(list);

        savedInstanceState.putString("ListaGraczy", json);
        savedInstanceState.putInt("MaksymalnyPoziom", MaxPlayerLevel);
        savedInstanceState.putInt("MinimalnyPoziom", MinLevel);
    }

    //Odebranie danych z innych Activity
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data)
    {
        //Aktualizacja poziomu z Kill-O-Meter'a
        if(requestCode == 1)
        {
            MaxPlayerLevel = data.getIntExtra("resultMaxLevel", getResources().getInteger(R.integer.deafult_rules));
            MinLevel = data.getIntExtra("resultMinLevel", getResources().getInteger(R.integer.deafult_min_level));
            int resultPosition = data.getIntExtra("resultPosition", 0);
            int resultLevel = data.getIntExtra("resultLevel", MinLevel);
            String json = data.getStringExtra("resultList");
            Type listType = new TypeToken<ArrayList<Player>>(){}.getType();
            list = new Gson().fromJson(json, listType);
            if(resultCode == RESULT_OK)
            {
                list.get(resultPosition).level = resultLevel;
                setPlayerAdapter();
            }
        }
        //Sprawdzenie poziomów po aktualizacji maksymalnego poziomu
        else if(requestCode == 2)
        {
            editMode = data.getBooleanExtra("resultEditMode", true);
            MaxPlayerLevel = data.getIntExtra("resultMaxLevel", getResources().getInteger(R.integer.deafult_rules));
            MinLevel = data.getIntExtra("resultMinLevel", getResources().getInteger(R.integer.deafult_min_level));
            String json = data.getStringExtra("resultList");
            Type listType = new TypeToken<ArrayList<Player>>(){}.getType();
            list = new Gson().fromJson(json, listType);
            if(resultCode == RESULT_OK)
                setPlayerAdapter();
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    //Zapisywanie listy gier do SharedPreferences
    public void saveGameListInSharedPreferences(List<Game> list)
    {
        Gson gsonGame = new Gson();
        String jsonGame = gsonGame.toJson(list);
        SharedPreferences.Editor editor = getSharedPreferences(SharedPrefs, MODE_PRIVATE).edit();
        editor.putString("ListaGierPrefs", jsonGame);
        editor.commit();
    }

    //Zserializuj listę graczy do jsona
    public String serializePlayerList(List<Player> list)
    {
        Gson gsonList = new Gson();
        String jsonList = gsonList.toJson(list);
        return jsonList;
    }

    //Stwórz domyślną grę
    public void createDefaultGame(List<Player> list)
    {
        list.add(new Player("Gracz 1", MinLevel));
        list.add(new Player("Gracz 2", MinLevel));
        list.add(new Player("Gracz 3", MinLevel));

        MaxPlayerLevel = getResources().getInteger(R.integer.deafult_rules);
        MinLevel = getResources().getInteger(R.integer.deafult_min_level);
    }
}

File: MunchkinLevelCounter/app/src/main/java/com/gmail/miloszwasacz/munchkinlevelcounter/MainActivity.kt
package com.gmail.miloszwasacz.munchkinlevelcounter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    internal lateinit var adapter: PlayerAdapter
    internal lateinit var game: Game
    internal var gameIndex: Int? = null
    internal var sharedPrefsName = "com.gmail.miloszwasacz.munchkinlevelcounter.prefs"
    internal lateinit var gameList: ArrayList<Game>
    internal var editMode: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        supportActionBar!!.setDisplayHomeAsUpEnabled(false)
        supportActionBar!!.title = "Licznik"

        gameList = getGameListFromSharedPreferences() ?: ArrayList<Game>()
        game = Game("", serializePlayerList(ArrayList<Player>()))

        //Przywracanie stanu poprzedniego listy graczy
        if(savedInstanceState != null){
            val jsonGame = savedInstanceState.getString("Gra")
            val gameType = object : TypeToken<Game>() {}.type
            game = Gson().fromJson<Game>(jsonGame, gameType)

            val jsonGameList = savedInstanceState.getString("ListaGier")
            val gameListType = object : TypeToken<ArrayList<Game>>() {}.type
            gameList = Gson().fromJson<ArrayList<Game>>(jsonGameList, gameListType)

            val jsonIndex = savedInstanceState.getString("Index")
            val indexType = object : TypeToken<Int>() {}.type
            gameIndex = Gson().fromJson<Int>(jsonIndex, indexType)
        }
        else {
            if(gameList.isEmpty()) {
                createNewGame(gameList)
            }
            else {
                loadGame(gameList)
            }
        }
        setPlayerAdapter(game)

        //Zmiana trybu guzika: edycja/dodawanie graczy
        floatingActionButton.setOnClickListener {
            if (adapter.editMode) {
                val list = extractPlayerListFromGame(game)
                list.add(Player("Nowy gracz", game.minLevel))
                insertPlayerListIntoGame(list, game)
                setPlayerAdapter(game)
            } else
                changeEditMode(R.drawable.ic_baseline_add_white_24dp, true, "Edytuj graczy")
        }
    }

    //RecyclerView i PlayerAdapter
    fun setPlayerAdapter(game: Game) {
        recycler_view.setHasFixedSize(true)
        recycler_view.layoutManager = LinearLayoutManager(this)
        val list: ArrayList<Player> = extractPlayerListFromGame(game)
        adapter = PlayerAdapter(list)
        adapter.editMode = editMode

        //Sprawdzenie czy poziomy graczy są w dozwolonym zakresie
        for (element in list) {
            if (element.level > game.maxLevel)
                element.level = game.maxLevel
            else if (element.level < game.minLevel)
                element.level = game.minLevel
        }

        //Obsługa kontrolek
        adapter.setOnItemClickListener(object : PlayerAdapter.OnItemClickListener {
            //Edycja poszczególnego gracza
            override fun onItemClick(position: Int) {
                val frameLayout = layoutInflater.inflate(R.layout.player_dialog, null, false) as FrameLayout
                val editText = frameLayout.findViewById<EditText>(R.id.editText)
                editText.setText(list[position].name)

                AlertDialog.Builder(this@MainActivity)
                        .setTitle("Edytuj gracza")
                        .setPositiveButton("Ok") { dialog, which ->
                            list[position].name = editText.text.toString()
                            adapter.notifyItemChanged(position)
                        }
                        .setNegativeButton("Anuluj", null)
                        .setNeutralButton("Usuń") { dialog, which ->
                            list.removeAt(position)
                            adapter.notifyItemRemoved(position)
                        }
                        .setView(frameLayout)
                        .create()
                        .show()

                insertPlayerListIntoGame(list, game)
            }

            //Zwiększenie poziomu gracza
            override fun onAddClick(position: Int) {
                if (list[position].level < game.maxLevel) {
                    list[position].level++
                    adapter.notifyItemChanged(position)
                    insertPlayerListIntoGame(list, game)
                }
            }

            //Zmniejszenie poziomu gracza
            override fun onRemoveClick(position: Int) {
                if (list[position].level > game.minLevel) {
                    list[position].level--
                    adapter.notifyItemChanged(position)
                    insertPlayerListIntoGame(list, game)
                }
            }

            //Wejście w tryb Kill-O-Meter
            override fun onFightClick(position: Int) {
                insertPlayerListIntoGame(list, game)
                val intent = Intent(this@MainActivity, KillOMeterActivity::class.java)
                intent.putExtra("EXTRA_POSITION", position)
                intent.putExtra("EXTRA_GAME", Gson().toJson(game))
                startActivityForResult(intent, 1)
            }
        })

        /*
        //Wyświetlenie nazwy gry
        if(game.name == "")
            supportActionBar!!.title = "Licznik"
        else
            supportActionBar!!.title = game.name*/

        recycler_view.adapter = adapter
    }

    //Włączanie/wyłączanie trybu edycji
    fun changeEditMode(icon: Int, enabled: Boolean?, title: String) {
        supportActionBar!!.title = title
        floatingActionButton.setImageResource(icon)
        editMode = enabled!!
        adapter.editMode = enabled
        invalidateOptionsMenu()
        adapter.notifyDataSetChanged()
    }

    //Strzałeczka w tył (tryb edycji)
    override fun onBackPressed() {
        if (adapter.editMode)
            changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik")
        else
            super.onBackPressed()
    }

    //Tworzenie guzików na app barze
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menu.clear()
        menuInflater.inflate(R.menu.toolbar_menu, menu)
        val mSaveButton = menu.findItem(R.id.action_save)
        val mLoadButton = menu.findItem(R.id.action_folder)
        val mClearButton = menu.findItem(R.id.action_clear)
        val mSettingsButton = menu.findItem(R.id.navigation_settings)
        if (adapter.editMode) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            mSaveButton.isVisible = false
            mLoadButton.isVisible = false
            mClearButton.isVisible = false
            mSettingsButton.isVisible = true
        } else {
            supportActionBar!!.setDisplayHomeAsUpEnabled(false)
            mSaveButton.isVisible = true
            mLoadButton.isVisible = true
            mClearButton.isVisible = true
            mSettingsButton.isVisible = false
        }
        return super.onCreateOptionsMenu(menu)
    }

    //Obsługa guzików na app barze
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if(gameIndex != null)
            saveGame(game, gameList)
        gameList = getGameListFromSharedPreferences() ?: ArrayList<Game>()

        when (item.itemId) {
            //Tworzenie nowej rozgrywki
            R.id.action_save -> {
                createNewGame(gameList)
                /*
                val frameLayout = layoutInflater.inflate(R.layout.player_dialog, null, false) as FrameLayout
                val editText = frameLayout.findViewById<View>(R.id.editText) as EditText

                AlertDialog.Builder(this@MainActivity)
                        .setTitle("Dodaj grę")
                        .setPositiveButton("Ok") { dialog, which ->
                            if (gameList == null)
                                gameList = ArrayList()

                            if (editText.text.toString() == "")
                                editText.setText("Gra " + (gameList!!.size + 1))

                            val playerList = ArrayList<Player>()
                            playerList.add(Player("Gracz 1", resources.getInteger(R.integer.default_min_level)))
                            playerList.add(Player("Gracz 2", resources.getInteger(R.integer.default_min_level)))
                            playerList.add(Player("Gracz 3", resources.getInteger(R.integer.default_min_level)))
                            game = Game(editText.text.toString(), serializePlayerList(playerList), resources.getInteger(R.integer.default_rules), resources.getInteger(R.integer.default_min_level))
                            gameList!!.add(game)
                            saveGameListInSharedPreferences(gameList)

                            changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik")
                            Toast.makeText(this@MainActivity, "Zapisano rozgrywkę", Toast.LENGTH_SHORT).show()
                        }
                        .setNegativeButton("Anuluj", null)
                        .setView(frameLayout)
                        .create()
                        .show()*/

                return true
            }

            //Wczytanie ostatniej rozgrywki
            R.id.action_folder -> {
                loadGame(gameList)
                /*if (gameList == null)
                    gameList = ArrayList()
                val lista = ArrayList<Player>()
                //createDefaultGame(lista)
                gameList!!.add(0, Game("default", serializePlayerList(lista), maxPlayerLevel, minLevel))

                val nameArray = arrayOfNulls<String>(gameList!!.size)
                for (i in gameList!!.indices) {
                    nameArray[i] = gameList!![i].name
                }

                AlertDialog.Builder(this@MainActivity)
                        .setTitle("Wczytaj rozgrywkę")
                        .setItems(nameArray) { dialog, which ->
                            maxPlayerLevel = gameList!![which].maxLevel
                            minLevel = gameList!![which].minLevel
                            val jsonGame = gameList!![which].content
                            val listType = object : TypeToken<ArrayList<Player>>() {}.type
                            list = Gson().fromJson<ArrayList<Player>>(jsonGame, listType)
                            setPlayerAdapter()
                            changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik")

                            Toast.makeText(this@MainActivity, "Wczytano rozgrywkę", Toast.LENGTH_SHORT).show()
                        }
                        .create()
                        .show()*/

                return true
            }

            //Usunięcie gry z listy
            R.id.action_clear -> {
                deleteGame(gameList)
                /*
                if (gameList == null)
                    Toast.makeText(this@MainActivity, "Brak zapisanych rozgrywek", Toast.LENGTH_SHORT).show()
                else {
                    val nameArrayDelete = arrayOfNulls<String>(gameList!!.size)

                    for (i in gameList!!.indices) {
                        nameArrayDelete[i] = gameList!![i].name
                    }

                    AlertDialog.Builder(this@MainActivity)
                            .setTitle("Usuń rozgrywkę")
                            .setItems(nameArrayDelete) { dialog, which ->
                                gameList!!.removeAt(which)
                                saveGameListInSharedPreferences(gameList)

                                changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik")
                                Toast.makeText(this@MainActivity, "Usunięto rozgrywkę", Toast.LENGTH_SHORT).show()
                            }
                            .create()
                            .show()
                }*/
                return true
            }

            //Otwarcie ustawień
            R.id.navigation_settings -> {
                val intent = Intent(this@MainActivity, SettingsActivity::class.java)
                intent.putExtra("EXTRA_MAX_LEVEL", game.maxLevel)
                intent.putExtra("EXTRA_MIN_LEVEL", game.minLevel)
                intent.putExtra("EXTRA_EDIT_MODE", adapter.editMode)
                startActivityForResult(intent, 2)
                return true
            }

            //Strzałeczka "back"
            android.R.id.home -> {
                onBackPressed()
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    //Zapisywanie stanu listy graczy
    public override fun onSaveInstanceState(savedInstanceState: Bundle) {
        super.onSaveInstanceState(savedInstanceState)
        val game = Gson().toJson(game)
        val gameList = Gson().toJson(gameList)
        val index = Gson().toJson(gameIndex)
        savedInstanceState.putString("Gra", game)
        savedInstanceState.putString("ListaGier", gameList)
        savedInstanceState.putString("Index", index)
    }

    //Odebranie danych z innych Activity
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        //Aktualizacja poziomu z Kill-O-Meter'a
        if (requestCode == 1) {
            val json = data!!.getStringExtra("resultGame")
            val listType = object : TypeToken<Game>() {}.type
            game = Gson().fromJson<Game>(json, listType)
            if (resultCode == Activity.RESULT_OK) {
                setPlayerAdapter(game)
            }
        }
        //Sprawdzenie poziomów po aktualizacji maksymalnego poziomu
        else if (requestCode == 2) {
            editMode = data!!.getBooleanExtra("resultEditMode", true)
            game.maxLevel= data.getIntExtra("resultMaxLevel", resources.getInteger(R.integer.default_rules))
            game.minLevel = data.getIntExtra("resultMinLevel", resources.getInteger(R.integer.default_min_level))
            if (resultCode == Activity.RESULT_OK)
                setPlayerAdapter(game)
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    //Zapisywanie listy gier do SharedPreferences
    fun saveGameListInSharedPreferences(list: ArrayList<Game>) {
        val jsonGame = Gson().toJson(list)
        val editor = getSharedPreferences(sharedPrefsName, Context.MODE_PRIVATE).edit()
        editor.putString("ListaGierPrefs", jsonGame)
        editor.commit()
    }

    //Wczytanie listy gier z SharedPreferences
    fun getGameListFromSharedPreferences(): ArrayList<Game>? {
        val prefs = getSharedPreferences(sharedPrefsName, Context.MODE_PRIVATE)
        val listaGier: String? = prefs.getString("ListaGierPrefs", null)
        val listType = object : TypeToken<ArrayList<Game>>() {}.type
        return when(listaGier) {
            null -> null
            else -> Gson().fromJson<ArrayList<Game>>(listaGier, listType)
        }
    }

    //Serializowanie listy graczy do jsona
    fun serializePlayerList(list: ArrayList<Player>): String = Gson().toJson(list)

    //Tworzenie domyślnej gry
    fun createDefaultGame(name: String): Game {
        val playerList = ArrayList<Player>()
        playerList.add(Player("Gracz 1", resources.getInteger(R.integer.default_min_level)))
        playerList.add(Player("Gracz 2", resources.getInteger(R.integer.default_min_level)))
        playerList.add(Player("Gracz 3", resources.getInteger(R.integer.default_min_level)))
        return Game(name, serializePlayerList(playerList))
    }

    //Wczytanie listy graczy z gry
    fun extractPlayerListFromGame(game: Game): ArrayList<Player> {
        val json = game.content
        val listType = object : TypeToken<ArrayList<Player>>() {}.type
        return Gson().fromJson<ArrayList<Player>>(json, listType)
    }

    //Zapisywanie listy graczy w grze
    fun insertPlayerListIntoGame(list: ArrayList<Player>, game: Game) {
        val json = serializePlayerList(list)
        game.content = json
    }

    //Tworzenie nowej gry
    fun createNewGame(inputList: ArrayList<Game>) {
        val list = ArrayList<Game>()
        list.addAll(inputList)
        val frameLayout = layoutInflater.inflate(R.layout.player_dialog, null, false) as FrameLayout
        val editText = frameLayout.findViewById<View>(R.id.editText) as EditText
        AlertDialog.Builder(this@MainActivity)
                .setTitle("Stwórz grę")
                .setPositiveButton("Ok") { dialog, which ->
                    if (editText.text.toString() == "")
                        editText.setText("Gra " + (list.size + 1))
                    game = createDefaultGame(editText.text.toString())
                    list.add(game)
                    gameIndex = (list.size - 1)
                    setPlayerAdapter(game)
                    gameList = list
                    saveGame(game, gameList)

                    changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik")
                }
                .setNegativeButton("Anuluj", null)
                .setView(frameLayout)
                .create()
                .show()
    }

    //Wczytanie gry z listy
    fun loadGame(list: ArrayList<Game>) {
        if(list.isNotEmpty()) {
            val nameArray = arrayOfNulls<String>(list.size)
            for (i in list.indices)
                nameArray[i] = list[i].name

            //game = list[0]
            //gameIndex = 0

            AlertDialog.Builder(this@MainActivity)
                    .setTitle("Wczytaj rozgrywkę")
                    .setItems(nameArray) { dialog, which ->
                        game = list[which]
                        gameIndex = which
                        setPlayerAdapter(game)
                        changeEditMode(R.drawable.ic_baseline_edit_white_24dp, false, "Licznik")
                        Toast.makeText(this@MainActivity, "Wczytano rozgrywkę", Toast.LENGTH_SHORT).show()
                        gameList = list
                        saveGame(game, gameList)
                    }
                    .create()
                    .show()
        }
        else
            Toast.makeText(this@MainActivity, "Brak zapisanych rozgrywek", Toast.LENGTH_SHORT).show()
    }

    //Usuwanie aktywnej gry
    fun deleteGame(gameList: ArrayList<Game>) {
        if(gameIndex != null /*&& game.name != ""*/)
        {
            AlertDialog.Builder(this@MainActivity)
                    .setTitle("Usuń grę")
                    .setPositiveButton("Tak") { dialog, which ->
                        gameList.removeAt(gameIndex!!)
                        saveGameListInSharedPreferences(gameList)
                        setPlayerAdapter(Game("", serializePlayerList(ArrayList<Player>())))
                        gameIndex = null
                        if (gameList.isEmpty())
                            createNewGame(gameList)
                        else
                            loadGame(gameList)
                    }
                    .setNeutralButton("Nie", null)
                    .create()
                    .show()
        }
        else
            Toast.makeText(this@MainActivity, "Brak aktywnej rozgrywki", Toast.LENGTH_SHORT).show()
    }

    //Zapisywanie aktywnej gry i listy gier
    fun saveGame(game: Game, gameList: ArrayList<Game>) {
        gameList[gameIndex!!] = game
        saveGameListInSharedPreferences(gameList)
    }

    //Zapisywanie gry podczas wyjścia z aplikacji
    override fun onPause() {
        if(gameIndex != null)
            saveGame(game, gameList)
        Toast.makeText(this@MainActivity, "Zapisano rozgrywkę", Toast.LENGTH_SHORT).show()
        super.onPause()
    }
}
